Replace Karma + Mocha with Jest

eddyerburgh opened the pull request:
- Replace Karma and Mocha with Jest

I've added alias resolving when running tests with [babel-plugin-module-resolver](https://github.com/tleunen/babel-plugin-module-resolver)

I've omitted src/main.js and src/router/index.js from test coverage because I don't think we should encourage testing of those files

This implements part of this issue - https://github.com/vuejs-templates/webpack/issues/817

The next part is to add vue-test-utils, which I'll do when it's published.

johnsardine commented:
I support this PR, however the coverage report is very poor.

![screenshot 2017-08-01 08 24 59](https://user-images.githubusercontent.com/679335/28813848-f0389edc-7692-11e7-896c-c6ac7513dd12.png)

I also tried to implement Jest here [johnsardine/vue-jest](https://github.com/johnsardine/vue-jest) and the coverage is a little better, however not perfect. I would like you to take a look, if you can.

I have a fork of [jest-vue-preprocessor](https://github.com/johnsardine/jest-vue-preprocessor) which provides babel sourcemap to jest-transform.

I think the coverage should be fixed before this is accepted.

On my test, the coverage is off by one line, and this affects the sourcemap on the browser as well. I think the issue must come from the map that babel generates from the extracted .vue file script.

I have put some hours into this without any progress other than the one I report here. I'll have to continue trying.

eddyerburgh commented:
Yeah you're right, we should get correct source mapping before merging.

As you said, jest-vue-preprocessor doesn't return a map or have any logic to produce one. I'm going to play around with the vueify code and jest.

johnsardine commented:
@eddyerburgh Take a look at my fork of [jest-vue-preprocessor](https://github.com/johnsardine/jest-vue-preprocessor), I've added (not in the best way, I admit) support for sourcemaps from babel. The thing is, the sourcemap is alright, but the coverage report is one line off, and it affects the source map on the web as well

rit commented:
Is jest's license the same as the React's license? If it is, what is the implication?

matt-oconnell commented:
@rit Unfortunately, it is. Although there is an open issue to discuss removing it [here](https://github.com/facebook/jest/issues/4109).

michalsnik commented:
Great idea @eddyerburgh I was about to prepare the same PR :D It's a great step forward, `jest` simplifies whole unit testing process drastically 🚀 

This so called React's license main purpose is to protect themselves from patent trolls that are suing them for crazy amounts of money each month. Dan explained it recently on twitter and I don't think we should consider it a blocker..

eddyerburgh commented:
Before we merge we need to create a vue jest preprocessor that adds correct source mapping. There are plans for an official preprocessor, but I'm not sure how long it will be until it is published

eddyerburgh commented:
@johnsardine I've created a vue jest preprocessor that adds source maps - https://github.com/eddyerburgh/jest-vue. I tried to add to vire's existing library, but found it easier to create a fresh repo.

In light of the patent drama, are we still happy to go ahead with replacing Karma and mocha with Jest?

johnsardine commented:
@eddyerburgh This is great, thanks a lot. I was testing it in my [vue-jest](https://github.com/johnsardine/vue-jest) sample project and so far seems to be working great. Will update a bigger project this week, if I encounter any issues I'll give you more feedback.

eddyerburgh commented:
Are people happy for me to fix conflicts and get this branch merged? @LinusBorg

LinusBorg commented:
yes!

eddyerburgh commented:
conflicts fixed 👷

eddyerburgh commented:
/ping @collaborators

LinusBorg commented:
Great work!

A thought though: I'm not sure if it's a good idea to make a hard switch to jest without any fallback for people that are used to the "old" workflow. Essentially, we would force everyone to either use Jest or scrap together a config from their old projects.

But we should do the switch sometime, so... I think I want to check out if it's achievable to support both, for a while, through additional/other options in the meta.js file. We could preselect Jest and make people aware that we will deprecate the karma/mocha config soon.

eddyerburgh commented:
Ok, makes sense 😄  
If I add Jest as the default unit tests option, and include karma and mocha options, will you be happy to merge?

LinusBorg commented:
Absolutely!

Aladdin-ADD commented on template/build/webpack.test.conf.js:
```diff
@@ -1,12 +1,12 @@
-const webpack = require('webpack')
-const merge = require('webpack-merge')
-const baseWebpackConfig = require('./webpack.base.conf')
+var utils = require('./utils')
```
is there any reason to use `var`, not `const`?

eddyerburgh replied:
I though it was used before my changes, but it looks like its switched to use strict and const. I'll change it back 👍

chlab commented:
@eddyerburgh thanks for this! Was planning on doing the same! I just tried this out and am running into a few issues. This is on a fresh project setup with `vue init eddyerburgh/webpack` so there's barely any custom code.

**1.**
There seems to be an issue with loading modules:

```
Cannot find module './src/vuex' from 'ParticipantsList.vue'
```
The line in ParticipantsList.vue: `import { mapState } from 'vuex'`

A similar case:
```
Cannot find module './src/components/Participant' from 'ParticipantsList.vue'
```
The import statement: `import Participant from '@/components/Participant'`

**2.**
When running `npm run unit` I also get these errors regarding code coverage:

```
Running coverage on untested files...
    Failed to collect coverage from src/App.vue
    ERROR: TypeError: Cannot read property 'fileCoverage' of undefined
    STACK: TypeError: Cannot read property 'fileCoverage' of undefined
    at Instrumenter.instrumentSync (node_modules/istanbul-lib-instrument/dist/instrumenter.js:147:39)
```

**3.**
To debug this I setup another quick test project with these options:

```
? Project name test-setup
? Project description A Vue.js project
? Author The Author
? Vue build standalone
? Install vue-router? No
? Use ESLint to lint your code? No
? Setup unit tests Yes
? Pick a test runner jest
? Setup e2e tests with Nightwatch? No
```

This results in an invalid package.json:

```
"unit": "jest test/unit/specs --coverage",
,
"test": "npm run unit"    "build": "node build/build.js"
```
extra comma between the two script defintions and a missing comma after the `test` script definition.

Thanks again for your work!

gkatsanos commented:
lets please try to merge this!

eddyerburgh commented:
I've updated the PR and fixed the issues  @chlab highlighted.

I'm banging my head against the wall trying to get the test to run. The yes command doesn't trigger the vue-cli build, and the test hangs. I've tried lots of different ways of passing input to the command, but no luck.

Does anyone have any ideas?

LinusBorg commented:
@eddyerburgh running those tests is indeed a small nightmare. I was having the same problems while trying to switch the circleCI config to v2 - the questionnaire just at the first question. 

So don't worry too much about it, it seems to magically work on circleci v1, for whatever reason. 

I'm stroll looking for ways to automate the fill out of the questionnaire in any other way, so open for suggestions as well.

sinaa commented:
Sorry if I'm jumping into this issue, but the following should work to get the vue cli to run (not sure if you're already following the same approach):

```bash
echo 'echo' > fill.sh; yes | xargs sh fill.sh | vue init webpack my-project; rm -f fill.sh
```
For me, `yes` doesn't trigger the vue-cli build, but the "hack" above works fine.

LinusBorg commented:
@sinaa thanks, will give that a spin!

eddyerburgh commented:
I think this PR is ready to merged. If anyone would like to test and give feedback it would be really useful 😀

chlab commented:
I justed re-inited my project with the updated template and it looks like the issues I encountered the first time around are all solved 👏  Thanks @eddyerburgh

LinusBorg commented on docs/unit.md:
```diff
@@ -2,32 +2,8 @@
 
 An overview of the tools used by this boilerplate for unit testing:
 
-- [Karma](https://karma-runner.github.io/): the test runner that launches browsers, runs the tests and reports the results to us.
```
@eddyerburgh Since we are leaving the karma setup available, I don't think we should remove this docs section. 

We should rather leave that docs in, but prepend the page with new infos about jest.

LinusBorg commented on template/package.json:
```diff
@@ -99,6 +114,32 @@
     "webpack-hot-middleware": "^2.18.2",
     "webpack-merge": "^4.1.0"
   },
+  {{#unit}}
```
shouldn't that be `{{#if_eq runner "jest"}}`?

LinusBorg commented on template/src/router/index.js:
```diff
@@ -8,7 +8,11 @@ export default new Router({
+      name: 'HelloWorld',
+=======
       name: 'Hello',
+>>>>>>> 89c6fb18d4640afd6a5e89106b2493b4465d36a3
```
You seem to have some leftovers from an older merge conflict...

LinusBorg commented on template/test/unit/.eslintrc:
```diff
@@ -1,9 +1,5 @@
 {
   "env": {
-    "mocha": true
```
since we are leaving the karma/mocha setup in place, the eslint config should probably also leave this in.

It should be conditional:
```javascript
{
  "env": {{{#if_eq runner "karma"}}
    "mocha": true{{/if_eq}}{{#if_eq runner "jest"}}
    "jest"{{/if_eq}}
  }{{#if_eq runner "karma"}},
  "globals": {
    "expect": true,
    "sinon": true,
    "jest": true
  }{{/if_eq}}
}
```

LinusBorg commented on template/package.json:
```diff
@@ -5,7 +5,7 @@
   "private": true,
   "scripts": {
-    "dev": "webpack-dev-server --inline --progress --config build/webpack.dev.conf.js",
+    "dev": "node build/dev-server.js",
```
Hm, that merge doesn't look like it's supposed to ... the webpack-dev-server is new in the `develop`branch, while yours was still built on the old `dev-server` - so the former should stay in and the latter should go out.

But we might simply resolve this during / after the merge ..

LinusBorg commented:
Thanks for the hard work on this @eddyerburgh

gkatsanos commented:
what's missing from https://github.com/vuejs/vue-test-utils/ to consider it non-beta?

LinusBorg commented:
That's probably a question to ask the team working on that repository.
